bpf/lib/nodeport.h

```diff
@@ -1995,11 +1992,12 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __u32 *ifind
 		goto encap_redirect;
 #endif /* ENABLE_EGRESS_GATEWAY */
 
-	tuple.nexthdr = ip4->protocol;
-	tuple.daddr = ip4->daddr;
-	tuple.saddr = ip4->saddr;
+	ret = ipv4_extract_tuple(ctx, &tuple, &l4_off);
+	if (ret == DROP_CT_UNKNOWN_PROTO)
+		goto out;
+	if (IS_ERR(ret))
+		return ret;
```

Comment:
What's the rationale for this change? If it's just a performance optimization, note that it likely increases verifier complexity; I'm not sure the tradeoff is worth it.

Do we expect any user impact or is it just supposed to bypass useless code? The `IS_ERR(ret)` case in particular makes me think there might be some user impact.

Reply:
We shouldn't try to extract the L4 ports from a packet with unknown / unsupported L4 protocol.

Reply:
But we already check for the unknown protocol cases (i.e., not UDP, TCP, SCTP, or ICMP) in `ct_lookup{4,6}`. So why do we need to duplicate the check here?

Reply:
Ah, I see what you mean. In patch 4 we could/should tolerate the `DROP_CT_UNKNOWN_PROTO` and just continue. But for patch 5 we then strictly need that check - `ct_has_nodeport_egress_entry4()` expects the full 5-tuple, so we need to manually load the ports before even calling into `ct_lookup()`.

[long-term we can add an optimized ct_lookup_with_tuple() that doesn't extract the ports a second time. I suspect that would apply in other places too]

Reply:
Ah, I see. That makes sense. Could we have a small code comment to explain that?

Reply:
I ended up squashing those two patches together, and using a new LB helper that handles the port extraction internally.